engine/consensus/matching/engine.go

```diff
@@ -263,17 +272,57 @@ func (e *Engine) process(originID flow.Identifier, event interface{}) error {
 
 // onReceipt processes a new execution receipt.
 func (e *Engine) onReceipt(originID flow.Identifier, receipt *flow.ExecutionReceipt) error {
+	// when receiving a receipt, we might not be able to verify it if its previous result
+	// is unknown.  In this case, instead of dropping it, we store it in the pending receipts
+	// mempool, and process it later when its parent result has been received and processed.
+	// therefore, if a receipt is processed, we will check if it is the previous results of
+	// some pending receipts, if there are, then processing them one after another, as onReceipt
+	// is behind a lock
+	receiptID := receipt.ID()
+	resultID := receipt.ExecutionResult.ID()
+
+	processed, err := e.onCurrentReceipt(receipt)
+	if err != nil {
+		e.log.Error().
+			Err(err).
+			Hex("receipt_id", receiptID[:]).
+			Hex("result_id", resultID[:]).
+			Hex("origin_id", originID[:]).
+			Msg("could not process receipt")
+		return nil
+	}
+
+	if !processed {
+		return nil
+	}
+
+	childReceipts := e.pendingReceipts.ByPreviousResultID(resultID)
+	e.pendingReceipts.Rem(receiptID)
+
+	for _, childReceipt := range childReceipts {
+		// recursively processing the child receipts, since onReceipt
+		// is logging error internal already, we could ignore the returned
+		// error here
+		_ = e.onReceipt(childReceipt.ExecutorID, childReceipt)
+	}
+
+	return nil
```

Comment:
Returning `nil` in this function and ignoring errors goes against error handling guidelines that we are trying to follow. I believe that we can return nil only if that was invalid input or some other sentinel error. In case of exception or unexpected error it has to be returned to caller. This design is already implemented for new matching engine.

Reply:
I have different opinions for that, maybe we can discuss in 387 . So far this works, as the error [are always handled]https://github.com/onflow/flow-go/pull/407/files#diff-17e93dc346462f66b95ec3f31134583765b8fdc4eff5e3c03f276698a554997fR286).